File: 04/main.pl
use v5.38;
use Digest::MD5 "md5_hex";

open my $fh, "<", "input.txt" or die "Cannot open input file";
my $key = <$fh>;
chomp($key);

sub find_number($zeroes) {
    my $search_target = "0" x $zeroes;
    my $number = 0;
    while(1) {
        return $number if (md5_hex($key, $number) =~ /^$search_target/);
        $number++;
    }
}

say "Part1: " . find_number(5);
say "Part2: " . find_number(6);

File: 04/main.nim
import std/strutils
import checksums/md5

proc readkey(): string =
  let f = open("input.txt")
  defer: f.close()
  result = f.readLine()

proc find_number(key: string, zeroes: uint): int =
    result = 0
    let search_pattern = "0".repeat(zeroes)
    while true:
        if getMD5(key & result.intToStr()).startsWith(search_pattern):
            break
        inc result

let key = readkey()
echo "Part1: ", find_number(key, 5)
echo "Part2: ", find_number(key, 6)